### PR TITLE
Service backed by BoltDB

### PR DESCRIPTION
**Pull request by dlmiddlecote**
## Description

Added a service implementation backed by [BoltDB](https://github.com/boltdb/bolt), an embedded key-value database.

### CODE REVIEW

File: cmd/ooohh-api/main.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+			return errors.Wrap(err, "opening db")
+		}
+		defer db.Close()
+	
+		now := func() time.Time {
+			return time.Now()
+		}
+
+		s, err = service.NewService(db, now)
+	}
```
Comment (Sambigeara) [inline cmd/ooohh-api/main.go]: Out of interest, what's the reason for nesting this code block? Is it for the benefit of the `defer`? E.g. it will run db.Close() after `service.NewService(...)`?

File: pkg/service/service.go [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+	"github.com/boltdb/bolt"
+	"github.com/pkg/errors"
+	"github.com/segmentio/ksuid"
+	"github.com/vmihailenco/msgpack/v5"
+
+	"github.com/dlmiddlecote/ooohh"
+)
+
+var _ ooohh.Service = &service{}
```
Comment (Sambigeara) [inline pkg/service/service.go]: Why is this required? Somehow instantiating the interface or something?

File: pkg/service/service.go [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+}
+
+func NewService(db *bolt.DB, now func() time.Time) (*service, error) {
+
+	// Initialize top-level buckets.
+	txn, err := db.Begin(true)
+	if err != nil {
+		return nil, errors.Wrap(err, "beginning transaction")
+	}
+	defer txn.Rollback()
```
Comment (Sambigeara) [inline pkg/service/service.go:29]: Does the `defer`ed function call not get called if the block returns first?

**Review by Sambigeara — APPROVED** [review]

File: cmd/ooohh-api/main.go (hunk 1 above)
Comment (dlmiddlecote) [reply cmd/ooohh-api/main.go]: It’s mainly stylistic. The block creates a new scope, so anything declared inside the block won’t be available outside of it, unless it was declared outside (like the ooohh.Service). It’s just to group the component initialisation together. The defer is unaffected, and will be called after “run” exits.

File: pkg/service/service.go (hunk 2 above)
Comment (dlmiddlecote) [reply pkg/service/service.go]: It isn’t required. It’s actually a test. I should move it there. 

It’s essentially a check that `service` does indeed satisfy the Service interface. 

This article has more on it: https://dave.cheney.net/2017/06/11/go-without-package-scoped-variables

File: pkg/service/service.go (hunk 3 above)
Comment (dlmiddlecote) [reply pkg/service/service.go:29]: The defer will not be called if the function returns before the `defer` keyword. It will always be called once it’s been declared. I assume this is what you’re asking?

The other thing you could be asking is, why is it ok to always call this? Essentially, the last call is ‘commit’ if all is good, so we’ll be rolling back nothing.
